Attribution reporting: two new blogposts (#1736)
GoogleChrome/developer.chrome.com: site/en/docs/privacy-sandbox/attribution-reporting-event-guide/index.md

## Demo!

To generate real event-level reports and see them in your browser, try the [live demo](https://goo.gle/sppi-devrel-eventlevel).

Check out the [source code](https://github.com/GoogleChromeLabs/trust-safety-demo/tree/main/conversion-measurement)
for this demo. You can
[fork and customize](https://github.com/GoogleChromeLabs/trust-safety-demo/tree/main/conversion-measurement#fork-and-customize)
this demo code as needed.

{% Columns %}

{% Column %}
{% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/O59eUascDKlhrxR7SbJz.png", alt="ALT_TEXT_HERE", width="400", height="300" %}
{% endColumn %}

{% Column %}
{% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/u3Gfm4ebwIr9NREfq5eE.png", alt="ALT_TEXT_HERE", width="400", height="300" %}
{% endColumn %}

{% endColumns %}

## Enable the API in your browser

If you are developing locally or testing out a deployed demo app in your browser, it's enough to enable the API **in your browser**.

{% Details %}
{% DetailsSummary %}

How to enable the API in your browser

{% endDetailsSummary %}

Enable the API via a _flag_. A flag is a special setting in your browser.

> 1. Type or paste `chrome://flags/#enable-experimental-web-platform-features` in Chrome's URL bar. This works on mobile, too.

Reviewer: ```suggestion
1. Enter `chrome://flags/#enable-experimental-web-platform-features` in Chrome's URL bar. This works on mobile, too.
```

2. Click **Enabled.**
3. Restart Chrome.
4. Ensure that Privacy Sandbox features are turned **on** at `chrome://settings/privacySandbox`. If
   they're turned off, this would override your flag and disable the Attribution Reporting API.

<figure>
   {% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/bKfzMxqL4gpbDDw3NLDc.png", alt="ALT_TEXT_HERE", width="400", height="261" %}
   <figcaption>Turn on the flag to enable the API in your browser.</figcaption>
</figure>

{% endDetails %}

## Enable the API for end users (origin trial)

To enable the API for all your site's users⏤except those who have chosen to opt-out of Privacy Sandbox features⏤register your site for an _origin trial_. You can also register for an origin trial if your code is embedded in sites you don't own, as third-party code.

{% Aside 'key-term' %}
An _origin trial_ (sometimes called an _experiment_) is a way to test a new or experimental web platform feature, and give feedback to the web standards community on the feature.
Learn more in [Getting started with Chrome's origin trials](/blog/origin-trials/).
{% endAside %}

> ### Design your origin trial

Reviewer: Can you try making the `detailsSummary` an H3 element and see if it's included in the TOC? If yes, you don't need the duplicate subheading

Author: It does, but what I'm trying to do is to be able to deep-link to that section (e.g. if someone asks a question and the answer is in this section, I want to be able to send them a link to this section. Or if a reader wants to share a link to that specific section with someone else.)

Right now I made `detailsSummary` an H3 element, but I can't deep-link to it? There's no `#` I can click on its left, like for other titles.

Author: I tried both of these, they don't create the `#` for me:

```
{% Details %}
{% DetailsSummary 'h3' %}
How to design your origin trial
{% endDetailsSummary %}
```

```
{% Details %}
{% DetailsSummary 'h3' %}
### How to design your origin trial
{% endDetailsSummary %}
```

{% Details %}
{% DetailsSummary %}

How to design your origin trial

{% endDetailsSummary %}

Before you set up an origin trial, set time aside to design your experiment. Ask yourself and your team: _What is our goal with this experiment?_

If your goal is to understand the API mechanics, run your experiment as follows:

- Track conversions.
- See how you can assign different values to attribution trigger events (conversions).
- Look at the reports you're receiving.

If your goal is to see how well the API satisfies basic use cases, run your experiment as follows:

- Track conversions.
- Look at the count of reports you're receiving.
- [Recover the corrected conversion
  count](#recover-the-corrected-conversion-count). Optionally, if you want to try something more advanced, tweak the noise correction script.
  For example, try different groupings to see what sizes are necessary for the noise to be negligible.
- Compare the corrected count of conversions with source-of-truth data, that is cookie-based
  measurement data.

{% endDetails %}

> ### Set up an origin trial

Reviewer: Same as above, subheading seems redundant, especially if "How to set up an origin trial" can be a part of the TOC.

Author: (Same as above)

{% Details %}
{% DetailsSummary %}
How to set up an origin trial
{% endDetailsSummary %}

1. Head over to [Chrome Origin Trials: active
   trials](https://developers.chrome.com/origintrials/#/trials/active) and sign in.
2. Search for **Conversion Measurement** or **Attribution Reporting** and click **Register**.
3. Fill in the form:

>    - Third-party matching: if you're planning on using the API as a third-party—for
>      example, if you need to use the API in a script you wrote that is executed on origins you
>      don't own—check this. This is convenient if you need to test at scale across different
>      sites. Learn more in
>      [What are third-party origin trials?](/blog/third-party-origin-trials/)

Reviewer: ```suggestion
   - Third-party matching: Select this if you're planning on using the API as a third-party—for
     example, if you need to use the API in a script you wrote that is executed on origins you
     don't own. This is convenient if you need to test at scale across different
     sites. Learn more in
     [What are third-party origin trials](/blog/third-party-origin-trials/).
```

Reviewer: If it's a checkbox, this wording works.

>    - Subdomain matching: if you need your token to match all subdomains of the origin—for
>      example, if you need your `https://shop.example` token to be valid in
>      `https://shoes.shop.example` as well—check this.

Reviewer: ```suggestion
   - Subdomain matching: Select this if you need your token to match all subdomains of the origin—for
     example, if you need your `https://shop.example` token to be valid in
     `https://shoes.shop.example` as well.
```

>    - Usage restriction: keep **Standard Limit** (this should already be set as default).

Reviewer: ```suggestion
   - **Usage restriction**. Keep **Standard Limit** (this should already be set as default).
```

4. Click **Register**.

> 5. You get back an origin

Reviewer: ```suggestion
5. You'll get back an origin
```

>    trial token, which is essentially a long string. Add this token where needed via an HTTP header, meta tag or JavaScript.

Reviewer: ```suggestion
   trial token, which is a long string. Add this token where needed via an HTTP header, meta tag or JavaScript.
```

{% Aside 'gotchas' %}

- Even if an origin trial is set up for your site, the API will only be enabled for users
  that have **not opted out** of the Privacy Sandbox via `chrome://settings/privacySandbox`.
- If you're running an origin trial, make sure to follow [Troubleshooting origin
  trials](/blog/origin-trial-troubleshooting/). In particular if you're
  using the feature in an iframe, add the origin trial token to this iframe as well as in the top-level page.

{% endAside %}

{% endDetails %}

> ## Example code: basic usage

Reviewer: ```suggestion
## Example code: Basic usage
```

serving time, publisher site, and any other granular information needed by the adtech company at
ad-serving time.

Example value
: `"412444888111012"`

Limits
: Maximum 64 bits; must be an integer between 0 and 2^64-1. To avoid [issues with large numbers in JavaScript](https://2ality.com/2012/07/large-integers.html)
and ensure that the value is preserved, define `attributionsourceeventid` as a string, not as an
integer.

#### HTML attribute `attributiondestination` (required)

The [eTLD+1](https://web.dev/same-site-same-origin/#site) where a
conversion is expected for this ad.

Example value
: `"https://advertiser.example"`

Requirements
: Must be an [origin](https://web.dev/same-site-same-origin/#origin).
: Must use HTTPS (or `http://localhost`, or `http://127.0.0.1`).
: If `attributiondestination` is `https://advertiser.example`, conversions on both
`https://advertiser.example` and `https://shop.advertiser.example` will be attributed.
The same happens if `attributiondestination` is `https://shop.advertiser.example`: conversions on
both `https://advertiser.example` and `https://shop.advertiser.example` will be attributed.

{% Details %}
{% DetailsSummary %}

> Do's and Don'ts for `attributiondestination`

Reviewer: ```suggestion
Do's and don'ts for `attributiondestination`
```

{% endCompare %}

{% Compare 'better' %}

```text
attributiondestination="http://localhost:8082"
```

{% CompareCaption %} HTTP is fine for local origins. {% endCompareCaption %}
{% endCompare %}

{% Compare 'worse' %}

```text
attributiondestination="advertiser.example"
```

{% CompareCaption %}
This is not an [origin](https://web.dev/same-site-same-origin/#origin): the scheme (`https://`) is missing.
{% endCompareCaption %}
{% endCompare %}

{% Compare 'worse' %}

```text
attributiondestination="http://advertiser.example"
```

{% CompareCaption %}
This is an [origin](https://web.dev/same-site-same-origin/#origin), but the scheme is insecure (HTTP).
{% endCompareCaption %}
{% endCompare %}

{% endDetails %}

#### HTML attribute `attributionreportto`

The endpoint the browser will send reports to.

Example value
: `"https://adtech.example"`

Default value
: The top-level origin of the page where the `a` element is added or where the JavaScript API is called to register a click.

Requirements
: Must be an [origin](https://web.dev/same-site-same-origin/#origin).
: Must use HTTPS (or `http://localhost` or `http://127.0.0.1`).

{% Details %}
{% DetailsSummary %}

> Do's and Don'ts for `attributionreportto`

Reviewer: ```suggestion
Do's and don'ts for `attributionreportto`
```

attributionreportto="https://adtech.example/reports"
```

{% CompareCaption %}
This is not just an [origin](https://web.dev/same-site-same-origin/#origin), this is a full URL. This won't throw an error, but anything in the path after the origin will be ignored. Reports will be sent `https://adtech.example` and **not** to
`https://adtech.example/reports`.
{% endCompareCaption %}
{% endCompare %}

{% endDetails %}

#### HTML attribute `attributionexpiry`

In milliseconds, the duration for which conversions can be attributed to a source event (click)—from the moment the ad is clicked.

Example value
: `2592000000` = 30 days in milliseconds

Default value
: `864000000` = 10 days in milliseconds

Limits
: Minimum value: 2 days in milliseconds. Maximum value: 30 days in milliseconds.

#### HTML attribute `attributionsourcepriority`

A priority for the source event. Only the report that matches the source event with the highest priority will be sent.

Default value
: `0`

Example value
: `30`

Limits
: Minimum value: `0`. Maximum value: `100`.

#### HTML attribute `href`

The intended destination of the ad click.

Example value
: `"https://advertiser.example/blue-shoes"`

{% Aside %}

> **Why is a distinct attribute `attributiondestination` needed, on top of `href`? 🤔**

Reviewer: ```suggestion
Why is a distinct attribute `attributiondestination` needed, on top of `href`? 🤔
```

Reviewer: 2 much decoration :)

- The 302 redirect destination should keep the same
  [origin](https://web.dev/same-site-same-origin/#origin).
  For example, if the `/conversion` request above goes to `https://adtech.example`, a redirect such as `res.redirect( 302, 'https://reporter.adtech.example/.well-known/...'` will **not** trigger an attribution because `reporter.adtech` and `adtech` are not the same origin.
- The `/.well-known/` request tells the browser to send a report, but the report won't be sent immediately. See details in [Report scheduling](/docs/privacy-sandbox/attribution-reporting-event-introduction/#report-scheduling).
  {% endAside %}

### Allow the feature for cross-origin iframes

In order to use Attribution Reporting in **cross-origin iframes** to declare a click
or trigger an attribution, you need to enable the feature via the `attribution-reporting`
[Permissions-Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy).

```html
<iframe src="..." allow="attribution-reporting"></iframe>
```

This is not needed in a top-level document or in same-origin iframes.

> ## Example code: more advanced features

Reviewer: ```suggestion
## Example code: Advanced features
```

By default, the page reload will trigger a second conversion and you'll get _two_ reports.
This is not desirable, because the user really has made only _one_ purchase.

To avoid this, you can **deduplicate** conversions.

{% endDetails %}

#### Example code

Example use of `dedup-key` to deduplicate conversions for the same `orderId`:

Adtech's `server.js`:

```javascript/5,6
app.get('/conversion', (req, res) => {
   const conversionType = req.query['conversion-type']
   const triggerData = ...
   // Use the orderId as a deduplication key
   const orderId = req.query['order-id']
   const dedupKey=`${orderId}`
   const url = `/.well-known/attribution-reporting/trigger-attribution?trigger-data=${triggerData}&dedup-key=${dedupKey}`
   // Adtech orders the browser to schedule-send a report
   res.redirect(302, url)
})
```

{% Aside %}
You can combine deduplication keys and conversion priorities. In this case, the request URL will look like this:

```javascript
/.well-known/attribution-reporting/trigger-attribution?trigger-data=${triggerData}&priority=${priorityValue}&dedup-key=${dedupKey}
```

{% endAside %}

> ## Example code: reports

Reviewer: ```suggestion
## Example code: Reports
```

{% Aside 'gotchas' %}

Even though `localhost` and `127.0.0.1` are aliases, they're different origins. If you set `attributiondestination=http://localhost:3000`,
conversions happening on `http://127.0.0.1:3000` will **not** be registered, since the origins don't match.

{% endAside %}

## Troubleshooting in your browser

**If you've followed the [browser setup steps](#browser-setup)** but the API doesn't behave as expected, follow the troubleshooting tips below.

{% Aside %}
You can apply these techniques to debug an application
that's deployed or an application that's running locally.

If the tips below don't fix the issue you're encountering, join the
[mailing list for developers](https://groups.google.com/u/1/a/chromium.org/g/attribution-reporting-api-dev)
and ask your question.
{% endAside %}

### No clicks or trigger events are recorded

If you're not able to see any clicks or attribution trigger event on `chrome://conversion-internals` with your implementation, try the following:

- Check if the API is enabled for your browser: open `chrome://internals`. The **status** field should display **enabled**.

> - If the status **disabled**, follow the setup steps in [Browser setup](#browser-setup). Once you do that, the displayed status on `chrome://internals` should be **enabled**.

Reviewer: ```suggestion
- If the status is **disabled**, follow the setup steps in [Browser setup](#browser-setup). Once you do that, the displayed status on `chrome://internals` should be **enabled**.
```
